### ISSUE TITLE
signal only works in main thread during update bighorn rc1 -> rc2

### ISSUE THREAD

**Issue by jeroenmaelbrancke** [openvstorage/framework-extensions#85]
During the update from bighorn rc1 to rc2 the asd-manager returns a ValueError (signal only works in main thread).

It looks like threading cannot handle the signal using in the timeout parameter. https://github.com/openvstorage/framework-extensions/blob/develop/src/generic/decorators.py 

update.log
```
2018-03-14 16:28:23 40200 +0100 - LC1SRV0002 - 7448/139824783357696 - flask/decorators.py - new_function - 1881 - INFO - POST /update/install/<package_name> - Entering with [] {"package_name": "openvstorage-extensions"}
2018-03-14 16:28:23 40200 +0100 - LC1SRV0002 - 7448/139824783357696 - flask/decorators.py - new_function - 1883 - ERROR - Unexpected exception
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/ovs_extensions/api/decorators.py", line 111, in new_function
    return_data = f(*args, **kwargs)
  File "/opt/asd-manager/source/app/api.py", line 392, in update
    SDMUpdateController.update(package_name=package_name)
  File "/opt/asd-manager/source/controllers/update.py", line 110, in update
    cls._package_manager.install(package_name=package_name, client=cls._local_client)
  File "/usr/lib/python2.7/dist-packages/ovs_extensions/generic/decorators.py", line 32, in wrapper
    signal.signal(signal.SIGALRM, _handle_timeout)
ValueError: signal only works in main thread
```

asd-manager.log
```
2018-03-14 16:28:23 40200 +0100 - LC1SRV0002 - 7448/139824783357696 - flask/decorators.py - new_function - 1881 - INFO - POST /update/install/<package_name> - Entering with [] {"package_name": "openvstorage-extensions"}
2018-03-14 16:28:23 40200 +0100 - LC1SRV0002 - 7448/139824783357696 - flask/decorators.py - new_function - 1883 - ERROR - Unexpected exception
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/ovs_extensions/api/decorators.py", line 111, in new_function
    return_data = f(*args, **kwargs)
  File "/opt/asd-manager/source/app/api.py", line 392, in update
    SDMUpdateController.update(package_name=package_name)
  File "/opt/asd-manager/source/controllers/update.py", line 110, in update
    cls._package_manager.install(package_name=package_name, client=cls._local_client)
  File "/usr/lib/python2.7/dist-packages/ovs_extensions/generic/decorators.py", line 32, in wrapper
    signal.signal(signal.SIGALRM, _handle_timeout)
ValueError: signal only works in main thread
```

**Comment by JeffreyDevloo** [comment]
Flask is configured to handle all requests in a separate thread to ensure that multiple requests can be handled. This might be the issue.
On the other hand we should change the decorator to not use signal...

**Comment by JeffreyDevloo** [comment]
Relying on python-timeout-decorator to handle timing out now. Also account for multi-threading (which the signal did not)

**Comment by JeffreyDevloo** [comment]
Relying on the python-timeout decorator did not go too well. We had to account for the fact that the code could be called in a non-main thread.
This meant that multiprocessing was used, forking the process. The SSHClient instance was pickled and unpickled in the child process. The client, however, uses the same underlying socket, causing the child process to wait for eternity.
When the child process was terminated, the SSHClients connection was also closed in the main thread. My suspicion is that the deconstructor closes the underlying connection because the counter was reset when unpickling?

Another option for timing out is relying on the paramiko timeout which would close it's underlying channel
